2020-12-15-environment-variables-maintenance.md

```diff
@@ -0,0 +1,61 @@
+# Environment Variables Maintenance Proposal
+
+## The Problems
+
+### Documentation
+
+Comment by Ernesto:
+> I believe we need a policy or convention for environment variables. Right now in some projects we are using this policy: If you add a new env variable, you need to document it in both .env.sample and the RELEASE_NOTES file.
+
+### Propagation of the change
+
+It's also hard to maintain: when you pull updates you need to verify the RELEASE_NOTES and compare that with your env variables, you may miss that something changed or that something was added.
+
+### Different requirements for different environments
+
+Not all environment variables are required for development or testing. Currently we have no clear documentation regarding what is optional for development (sometimes a variable inside the .env.sample needs to be removed locally instead of leaving it blank for example).
+
+## Possible solutions
+
+Some suggestions by Ernesto and comments on Twitter were:
+
+### Envied
+
+URL: https://github.com/eval/envied
+
+Pros:
+- To use this gem we need to configure the variables in a `Envfile` (which is just ruby), and then run `ENVied.require` during the app boot.
+- We can add different requirements for different environments on a single file.
+- This can validate many formats.
+- Has a task to check the heroku config against the config file
+- Documentation can be added as comments on the Envfile
+
+Cons:
+- This does NOT support optional env variables (i.e.: variables that can be undefined but if present should have a specific format)
+
+### Prius
+
+URL: https://github.com/gocardless/prius
+
+Pros:
+- We can configure different variables for each environment
+- Supports optional variables
```

Comment:
@arielj Why do we care about optional variables? Just curious. Aren't all variables optional if they are not required? 😄 

Is the entire point to have defaults for these optional variables? If that is the case, wouldn't we just use the values set in `.env.sample`?

Reply:
Sorry, what I meant was support for format validation of optional variables, I'll update that. Not having it as required makes it optional, but with envied we can't add format validation, it's either required with some format or not required with any format. With prius we can validate the format if present but mark it as not required.

Also, one issue with the variables in .env.sample is that this file is added to repositories and it can't include sensitive defaults like API keys for example at least for open source projects (like dash's github api key)

Reply:
Not sure if we should care that much about format validation. How important is this?

In my experience, the main problem is: You forgot to add a new env variable that was recently added to the project. 

> Also, one issue with the variables in .env.sample is that this file is added to repositories and it can't include sensitive defaults like API keys for example at least for open source projects (like dash's github api key)

Right, we should never add sensitive values to the .sample files. I think that is an unspoken rule about those files. Usually those files include a fake value or a value that indicates where you should go and get the real value. 😄

Reply:
Regarding format, I had that issue a few times with Ombushop, in the env file we have a lot of variables with `SOME_VAR=<api key here>`, and some of those variables are expected to have a specific format and we have to go somewhere to find the right format to put a fake value there but that makes sense.